shorebird_code_push/lib/src/shorebird_updater.dart

```diff
@@ -0,0 +1,138 @@
+import 'package:shorebird_code_push/src/shorebird_updater_io.dart'
+    if (dart.library.js_interop) './shorebird_updater_web.dart';
+import 'package:shorebird_code_push/src/updater.dart';
+
+/// The reason a call to [ShorebirdUpdater.update] failed.
+enum UpdateFailureReason {
+  /// No update is available.
+  noUpdate,
+
+  /// The update failed because the patch failed to install.
+  installFailed,
+
+  /// The update failed because the patch could not be downloaded.
+  downloadFailed,
+
+  /// The update failed for an unknown reason.
+  unknown,
+}
+
+/// {@template read_patch_exception}
+/// An exception thrown by [ShorebirdUpdater.readCurrentPatch] and
+/// [ShorebirdUpdater.readNextPatch] when the read is unsuccessful.
+/// {@endtemplate}
+class ReadPatchException implements Exception {
+  /// {@macro update_exception}
+  const ReadPatchException({required this.message});
+
+  /// The human-readable error message.
+  final String message;
+}
+
+/// {@template update_exception}
+/// An exception thrown by [ShorebirdUpdater.update] when the update is
+/// unsuccessful.
+/// {@endtemplate}
+class UpdateException implements Exception {
+  /// {@macro update_exception}
+  const UpdateException({required this.message, required this.reason});
+
+  /// The human-readable error message.
+  final String message;
+
+  /// The reason the update failed.
+  final UpdateFailureReason reason;
+}
+
+/// Log message when the Shorebird updater is unavailable in the current
+/// environment.
+void logShorebirdEngineUnavailableMessage() {
+  // ignore: avoid_print
+  print('''
+-------------------------------------------------------------------------------
+The Shorebird Updater is unavailable in the current environment.
+-------------------------------------------------------------------------------
+This occurs when using pkg:shorebird_code_push in an app that does not
+contain the Shorebird Engine. Most commonly this is due to building with
+`flutter build` or `flutter run` instead of `shorebird release` or `shorebird preview`.
+It can also occur when running on an unsupported platform (e.g. web or desktop).
+''');
+}
+
+/// {@template patch}
+/// An object representing a single patch (over-the-air update).
+/// {@endtemplate}
+class Patch {
+  /// {@macro patch}
+  const Patch({required this.number});
+
+  /// The patch number.
+  final int number;
+}
+
+/// The current status of the app in terms of whether its up-to-date.
+enum UpdateStatus {
+  /// The app is up to date (e.g. running the latest patch.)
+  upToDate,
+
+  /// A new update is available for download.
+  outdated,
```

Comment:
Is this "downloadRequired"? to make it match restartRequired?  Or need(s)Download and needRestart?

Reply:
download isn't required ever right? Developers can choose not to download patches so I think outdated (consistent with upToDate) made more sense.

Reply:
sg